File: Data Structures/BinarySearchTree.js
/**
 * Tree Terminology
 * ================
 *
 * Root -> The Top Node.
 *
 * Child -> A node directly connected to another node, when moving away from the root.
 *
 * Parent -> The Converse notion of a child.
 *
 * Sibling -> The nodes with same parent.
 *
 * Leaf -> The node with no children.
 *
 * Edge -> The Connection between one Node and Other.
 */

/**
 * Time Complexity
 * ===============
 *
 * Insertion(Best & Average) => O(log N)
 *
 * Insertion(Worst) => O(N)
 *
 * Searching(Best & Average) => O(log N)
 *
 * Searching(Worst) => O(N)
 *
 */
class Node {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

class BinarySearchTree {
  constructor() {
    this.root = null;
  }
  insert(value) {
    let newNode = new Node(value);
    if (!this.root) {
      this.root = newNode;
      return this;
    } else {
      let current = this.root;
      while (true) {
        if (value === current.value) return null; // BSTs cannot have duplicates;
        if (value < current.value) {
          if (!current.left) {
            current.left = newNode;
            return this;
          } else {
            current = current.left;
          }
        } else if (value > current.value) {
          if (!current.right) {
            current.right = newNode;
            return this;
          } else {
            current = current.right;
          }
        }
      }
    }
  }
  find(value) {
    if (!this.root) return false;
    else if (this.root.value === value) return this.root;
    else {
      let current = this.root;
      while (true) {
        if (value === current.value) {
          return current;
        } else if (value < current.value) {
          if (!current.left) return false;
          else current = current.left;
        } else if (value > current.value) {
          if (!current.right) return false;
          else current = current.right;
        }
      }
    }
  }
}

let tree = new BinarySearchTree();
tree.insert(10);
tree.insert(5);
tree.insert(13);
tree.insert(2);
tree.insert(7);
tree.insert(11);
tree.insert(16);
// console.log(tree.insert(16)); // null
tree.insert(3);
// console.log(tree.root.left.value); // 5
// console.log(tree.root.left.left.value); // 2
// console.log(tree.root.left.right.value); // 7
// console.log(tree.root.right.value); // 13
// console.log(tree.root.right.left.value); // 11
// console.log(tree.root.right.right.value); // 16
// console.log(tree.root.left.left.right.value); // 3
console.log(tree.find(5)); // true
// console.log(tree.find(2)); // true
// console.log(tree.find(7)); // true
// console.log(tree.find(13)); // true
// console.log(tree.find(11)); // true
// console.log(tree.find(16)); // true
// console.log(tree.find(3)); // true
// console.log(tree.find(10)); // true
// console.log(tree.find(0)); // false

File: Data Structures/Practice_in_Python/BinarySearchTree.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, value):
        newNode = Node(value)
        if not self.root:
            self.root = newNode
            return self
        else:
            current = self.root
            while True:
                if current.value == value:
                    return None
                else:
                    if value < current.value:
                        if not current.left:
                            current.left = newNode
                            return self
                        else:
                            current = current.left
                    elif value > current.value:
                        if not current.right:
                            current.right = newNode
                            return self
                        else:
                            current = current.right

    def find(self, value):
        if self.root == None:
            return False
        elif value == self.root.value:
            return True
        else:
            current = self.root
            while True:
                if current.value == value:
                    return True
                elif value < current.value:
                    if current.left == None:
                        return False
                    else:
                        current = current.left
                elif value > current.value:
                    if current.right == None:
                        return False
                    else:
                        current = current.right

    def BFS(self):
        data = []
        queue = []
        node = self.root
        queue.append(node)
        while len(queue):
            node = queue.pop(0)
            data.append(node.value)
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        return data

    def DFSPreOrder(self):
        data = []
        current = self.root

        def traverse(node):
            data.append(node.value)
            if node.left:
                traverse(node.left)
            if node.right:
                traverse(node.right)

        traverse(current)
        return data

    def DFSPostOrder(self):
        data = []
        current = self.root

        def traverse(node):
            if node.left:
                traverse(node.left)
            if node.right:
                traverse(node.right)
            data.append(node.value)

        traverse(current)
        return data

    def DFSInOrder(self):
        data = []
        current = self.root

        def traverse(node):
            if node.left:
                traverse(node.left)
            data.append(node.value)
            if node.right:
                traverse(node.right)

        traverse(current)
        return data

    def height(self, root):
        if not root:
            return -1
        leftDepth = self.height(root.left)
        rightDepth = self.height(root.right)
        if leftDepth > rightDepth:
            return leftDepth + 1
        else:
            return rightDepth + 1


t = BinarySearchTree()
t.insert(5)
t.insert(3)
t.insert(8)
t.insert(1)
t.insert(4)
t.insert(9)
t.insert(6)
# print(t.find(10))
# print(t.find(2))
print(t.BFS())  # [5, 3, 8, 1, 4, 6, 9]
print(t.DFSPreOrder())  # [5, 3, 1, 4, 8, 6, 9]
print(t.DFSPostOrder())  # [1, 4, 3, 6, 9, 8, 5]
print(t.DFSInOrder())  # [1, 3, 4, 5, 6, 8, 9]
print(t.height(t.root))
# print(t.root.left.left.value)
